lib/spack/spack/solver/asp.py

```diff
@@ -470,6 +470,12 @@ def _compute_specs_from_answer_set(self):
                 self._concrete_specs.append(answer[node])
                 self._concrete_specs_by_input[input_spec] = answer[node]
             else:
+                msg = f"Unsatisfied input spec: {str(input_spec)}"
+                if candidate:
+                    msg += f"\n\tCandidate spec: {str(candidate)}"
+                else:
+                    msg += "\n\t(No candidate specs from solver)"
+                tty.debug(msg)
```

Comment:
(nitpick) I would avoid debugging messages here, as they might add a lot of entries in `when_possible` environments

Reply:
Fair enough: it was lazy of me to put this here, but in order to retain the information I'll need to update the structure of `_unsolved_specs` so that it can be retrieved at the later point (I want to show the user spec and associated candidate together).

Reply:
`Result.unsolved_specs` is now a list of tuples (input spec, result spec)

* This info is now dumped with the raised exception (which also means the user doesn't have to rerun)
* https://github.com/spack/spack/pull/42655#discussion_r1499379674 is also updated to output the associated candidate specs as well
* It seems to me we could do an early termination of `solve_in_rounds` if any of the values of `unsolved_specs` set (not `None`) although I haven't done that here